File: meshcall-ts/src/router-auth.ts
import type { MeshCallClient, CallOptions } from "./client.js";

export type RouterCredentials =
  | { readonly username: string; readonly password: string; readonly token?: never }
  | { readonly token: string; readonly username?: never; readonly password?: never };

export function authorizationHeader(auth: RouterCredentials): string {
  if (auth.token !== undefined) {
    if (!auth.token || /[\r\n]/u.test(auth.token) || auth.username !== undefined || auth.password !== undefined) {
      throw new TypeError("Set either a token or username/password");
    }
    return "Bearer " + auth.token;
  }
  if (!auth.username || !auth.password || /[:\r\n]/u.test(auth.username)) {
    throw new TypeError("A username and password are required");
  }
  return "Basic " + Buffer.from(auth.username + ":" + auth.password, "utf8").toString("base64");
}

export interface RouterScope {
  readonly service: string;
  readonly methods?: ReadonlyArray<string>;
  readonly register?: boolean;
}

export interface IssuedToken {
  readonly token: string;
  readonly token_id: string;
  readonly expires_at_unix_ms: number;
  readonly scopes: ReadonlyArray<RouterScope>;
}

export interface RouterIdentity {
  readonly username: string;
  readonly roles: ReadonlyArray<"client" | "server">;
  readonly register: ReadonlyArray<string>;
  readonly call: ReadonlyArray<string>;
  readonly expires_at_unix_ms: number | null;
}

export const ROUTER_AUTH_SERVICE = "meshcall.router.v1.AuthService";

export class RouterAuthClient {
  public constructor(private readonly client: MeshCallClient) {}

  public whoami(options?: CallOptions): Promise<RouterIdentity> {
    return this.client.unary(ROUTER_AUTH_SERVICE, "whoami", {}, options);
  }

  public issueToken(
    request: { readonly scopes: ReadonlyArray<RouterScope>; readonly ttl_seconds?: number },
    options?: CallOptions,
  ): Promise<IssuedToken> {
    return this.client.unary(ROUTER_AUTH_SERVICE, "issue_token", {
      ...request, ttl_seconds: request.ttl_seconds ?? 900,
    }, options);
  }

  public async revokeToken(tokenId: string, options?: CallOptions): Promise<boolean> {
    const result = await this.client.unary<{ token_id: string }, { revoked: boolean }>(
      ROUTER_AUTH_SERVICE, "revoke_token", { token_id: tokenId }, options,
    );
    return result.revoked;
  }
}

File: meshcall-py/src/meshcall/router_auth.py
from __future__ import annotations

import base64
from dataclasses import dataclass, field
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from meshcall.client import ClientBase

AUTH_SERVICE = "meshcall.router.v1.AuthService"


@dataclass(frozen=True)
class RouterCredentials:
    username: str | None = None
    password: str | None = field(default=None, repr=False)
    token: str | None = field(default=None, repr=False)

    def __post_init__(self) -> None:
        if self.token is not None:
            if not self.token or self.username is not None or self.password is not None:
                raise ValueError("Set either a token or username/password")
            if any(c in self.token for c in "\r\n"):
                raise ValueError("Invalid token")
        elif (
            not self.username
            or not self.password
            or any(c in self.username for c in ":\r\n")
        ):
            raise ValueError("A username and password are required")

    def authorization_header(self) -> str:
        if self.token is not None:
            return f"Bearer {self.token}"
        encoded = base64.b64encode(f"{self.username}:{self.password}".encode()).decode()
        return f"Basic {encoded}"


class RouterScope(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    service: str
    methods: list[str] = Field(default_factory=list)
    register_service: bool = Field(
        default=False, validation_alias="register", serialization_alias="register"
    )


class IssuedToken(BaseModel):
    token: str = Field(repr=False)
    token_id: str
    expires_at_unix_ms: int
    scopes: list[RouterScope]


class RouterIdentity(BaseModel):
    username: str
    roles: list[Literal["client", "server"]]
    register_services: list[str] = Field(alias="register")
    call: list[str]
    expires_at_unix_ms: int | None


class _Empty(BaseModel):
    pass


class _IssueRequest(BaseModel):
    ttl_seconds: int = Field(ge=1, le=3600)
    scopes: list[RouterScope]


class _RevokeRequest(BaseModel):
    token_id: str


class _RevokeResult(BaseModel):
    revoked: bool


class RouterAuthClient(ClientBase):
    """Typed client for the Go Router's built-in, unary management service."""

    async def whoami(self) -> RouterIdentity:
        return await self._unary(AUTH_SERVICE, "whoami", _Empty(), RouterIdentity)

    async def issue_token(
        self, *, scopes: list[RouterScope], ttl_seconds: int = 900
    ) -> IssuedToken:
        return await self._unary(
            AUTH_SERVICE,
            "issue_token",
            _IssueRequest(ttl_seconds=ttl_seconds, scopes=scopes),
            IssuedToken,
        )

    async def revoke_token(self, token_id: str) -> bool:
        result = await self._unary(
            AUTH_SERVICE,
            "revoke_token",
            _RevokeRequest(token_id=token_id),
            _RevokeResult,
        )
        return result.revoked
